runtime/reconcilers/metrics_view.go

```diff
@@ -119,11 +122,69 @@ func (r *MetricsViewReconciler) Reconcile(ctx context.Context, n *runtimev1.Reso
 		mv.State.Streaming = !hasInternalRef
 	}
 
+	// set cache controls
+	olap, release, err := r.C.AcquireOLAP(ctx, mv.Spec.Connector)
+	if err != nil {
+		return runtime.ReconcileResult{Err: fmt.Errorf("failed to get OLAP: %w", err)}
+	}
+	defer release()
+
+	validSpec := mv.State.ValidSpec
+	if validSpec != nil && validSpec.Cache != nil && validSpec.Cache.Enabled != nil && *validSpec.Cache.Enabled &&
+		mv.State.Streaming && validSpec.Cache.KeySql == "" && watermarkForMetricsView(validSpec) == "" {
+		return runtime.ReconcileResult{Err: fmt.Errorf("cache enabled for streaming metrics view %q but no `watermark` or `key_sql` provided", self.Meta.Name)}
+	}
+
 	// Update state. Even if the validation result is unchanged, we always update the state to ensure the state version is incremented.
 	err = r.C.UpdateState(ctx, self.Meta.Name, self)
 	if err != nil {
 		return runtime.ReconcileResult{Err: err}
 	}
+	// Override the cache control with our calculated cache control
+	if validSpec != nil {
+		validSpec.Cache = metricsViewCacheControl(validSpec, mv.State.Streaming, self.Meta.StateUpdatedOn, olap.Dialect())
+	}
 
 	return runtime.ReconcileResult{Err: validateErr}
 }
+
+func metricsViewCacheControl(spec *runtimev1.MetricsViewSpec, streaming bool, updatedOn *timestamppb.Timestamp, dialect drivers.Dialect) *runtimev1.MetricsViewSpec_Cache {
+	var enabled *bool
+	if spec.Cache != nil && spec.Cache.Enabled != nil {
+		enabled = spec.Cache.Enabled
+	} else {
+		enabled = boolPtr(!streaming)
+	}
+	cache := &runtimev1.MetricsViewSpec_Cache{
+		Enabled: enabled,
+	}
+	if spec.Cache != nil && spec.Cache.KeyTtlSeconds != 0 {
+		cache.KeyTtlSeconds = spec.Cache.KeyTtlSeconds
+	} else if streaming {
+		cache.KeyTtlSeconds = 60
+	}
+
+	if spec.Cache != nil && spec.Cache.KeySql != "" {
+		cache.KeySql = spec.Cache.KeySql
+	} else {
+		if streaming {
+			cache.KeySql = fmt.Sprintf("SELECT %s FROM %s", watermarkForMetricsView(spec), dialect.EscapeTable(spec.Database, spec.DatabaseSchema, spec.Table))
+		} else {
+			cache.KeySql = "SELECT " + fmt.Sprintf("'%d:%d'", updatedOn.GetSeconds(), updatedOn.GetNanos()/int32(time.Millisecond))
```

Comment:
May be the version is unnecessary for rill managed models because caches will anyways expire. We can also return an empty cache key ?

Reply:
See note above – if you move this logic to the resolver, you won't need to worry about creating a SQL statement that returns a constant

Reply:
This was more of a convenient way of generating `runtime.ResolverResult` using results of a SQL query. May be we can still keep the same behaviour but return a constant (not driven by `updatedOn`). Or do you have something else in mind ?

Reply:
Since the `key_sql` will require another DB roundtrip, it would be great only to do it for dynamic results (like `MAX(__time)`). When it's a constant, returning it immediately in Go (without the SQL roundtrip) would be nice.